File: 567.py
def checkInclusion(s1: str, s2: str):
    if (len(s1) > len(s2)):
        return False
    c1Counts = {}
    #count s1 and add to c1Counts
    for c in s1:
        if c in c1Counts:
            c1Counts[c] += 1
        else:
            c1Counts[c] = 1
    curr = max = 0
    c1Clones = c1Counts.copy()
    for c in s2:
        if c in c1Counts and c1Counts[c] > 0:
            curr += 1
            if curr > max:
                max = curr
            c1Counts[c] -= 1
        else:
            curr = 0
            c1Counts = c1Clones.copy()
    for i in range(len(s1)):
        if s2[i] in c1Counts and c1Counts[s2[i]] > 0:
            curr += 1
            if curr > max:
                max = curr
            c1Counts[s2[i]] -= 1
        else:
            curr = 0
            c1Counts = c1Clones.copy()
    return (max == len(s1))



s1 = input()
s2 = input()
print(checkInclusion(s1, s2))

File: 567.cpp
#include <iostream>
#include <unordered_map>
#include <string>

using namespace std;

bool checkInclusion(string s1, string s2) {
    unordered_map<char, int> c1Counts, c1CountsClone;
    int max, curr;
    max = curr = 0;
    for (char c : s1) {
        c1Counts[c]++;
    }
    c1CountsClone = c1Counts;
    for (int i = 0; i < s2.size(); i++) {
        char c = s2[i];
        if (c1Counts[c] > 0) {
            curr++;
            if (curr > max) {
                max = curr;
            }
            c1Counts[c]--;
        }
        else {
            curr = 0;
            c1Counts = c1CountsClone;
        }
    }
    for (int i = 0; i < s1.size(); i++) {
        char c = s2[i];
        if (c1Counts[c] > 0) {
            curr++;
            if (curr > max) {
                max = curr;
            }
            c1Counts[c]--;
        }
        else {
            curr = 0;
            c1Counts = c1CountsClone;
        }
    }
    if (max == s1.size()) {
        return true;
    }
    return false;
}

int main() {
    string s1, s2;
    cin >> s1 >> s2;
    printf(checkInclusion(s1, s2) ? "true" : "false");
}